### PR TITLE
Annotation placement analysis for Service, Pod, Ingress

### PR DESCRIPTION
**Pull request by esnible**
Mitigates https://github.com/istio/istio/issues/17686 by testing for annotation placement on pod, service, ingress, and a few other K8s objects known to the analysis system.

I flagged this a `work-in-progress` even though it is done because the included tests re-enable https://github.com/istio/istio/pull/17652 .  We shouldn't merge it unless we are ready to turn those tests back on.  I need those tests to prove to the reviewers that the code works.

### CODE REVIEW

File: galley/pkg/config/analysis/analyzers/annotations/annotations.go [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+)
+
+// K8sAnalyzer checks for misplayed and invalid Istio annotations in K8s resources
+type K8sAnalyzer struct{}
+
+var (
+	// Currently we don't have an Istio API that tells which Istio annotations apply
+	// to which K8s and custom resources.  This map is currently hand-crafted, but
+	// could be generated similarly to https://istio.io/docs/reference/config/annotations/
+	annotationAppliesTo = map[*annotation.Instance][]string{
```
Comment (nmittler) [inline galley/pkg/config/analysis/analyzers/annotations/annotations.go]: FYI @geeknoid might it make sense to merge annotation tools?

Comment (geeknoid) [reply galley/pkg/config/analysis/analyzers/annotations/annotations.go]: We're already covered...

This code should be pulling in the generated file in istio/api/annotations/annotations.gen.go. This includes the full current set of annotations. Sadly, it doesn't include the newly introduced resource type data that this analysis needs. I'll get this fixed, should be rolled out within a day.

Comment (geeknoid) [reply galley/pkg/config/analysis/analyzers/annotations/annotations.go]: https://github.com/istio/tools/pull/426

Comment (geeknoid) [reply galley/pkg/config/analysis/analyzers/annotations/annotations.go]: Once https://github.com/istio/api/pull/1121 merges, this PR can pull in the new .gen.go file to get all the necessary definitions.

**Comment by esnible** [comment]
@geeknoid @nmittler Thanks for reworking _api/annotation/annotations.gen.go_ to include the resource types that annotations apply to.  It is still unfortunate that there is no list of annotations, and no way to convert between the new annotation.ResourceTypes enumeration and Kubernetes 'Kind' strings.

@sushicw It is a pain that the analysis requires `--discovery` when run from the command line for my own files.  It also troubles me that this was requested by Lin after noticing people applying Pod annotations to Deployments, but we can't yet analyze deployments.  Any progress on those fronts?

**Comment by geeknoid** [comment]
@esnible The generated annotations.gen.go file does contain a list of all the annotations, doesn't it?

As for converting between the enumerated resource types and the associated k8s types, for now you'll need a map to do this.

**Comment by esnible** [comment]
> @esnible The generated annotations.gen.go file does contain a list of all the annotations, doesn't it?

The _annotations.gen.go_ has all the definitions, but no list.  They are just `var`.  Golang `reflect` can't extract the public vars from a package; I checked.

File: galley/pkg/config/analysis/analyzers/annotations/annotations.go [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+
+			// TODO: Check annotation.Deprecated.  Not implemented because no
+			// deprecations in the table have yet been deprecated!
+
+			// Note: currently we don't validate the value of the annotation,
+			// just key and placement.
+			// There is annotation value validation (for pod annotations) in the web hook at
+			// istio.io/istio/pkg/kube/inject/inject.go.  Rather than refactoring
+			// that code I intend to bring all of the web hook checks into this framework
+			// and checking here will be redundant.
```
Comment (sushicw) [inline galley/pkg/config/analysis/analyzers/annotations/annotations.go]: Link to an issue here?

File: galley/pkg/config/analysis/msg/messages.yaml [hunk 3]
```diff
@@ -37,6 +37,26 @@ messages:
       - name: detail
         type: string
 
+  - name: "MisplacedAnnotation"
+    code: IST0005
```
Comment (sushicw) [inline galley/pkg/config/analysis/msg/messages.yaml]: Is there a reason you're putting these in the reserved "internal" range? (See the comment at the top of `messages.yaml`). 

If not, the codes should be `IST0107` / `IST0108` (the next ones in sequence outside the reserved range).

Comment (esnible) [reply galley/pkg/config/analysis/msg/messages.yaml]: Done

File: galley/pkg/config/analysis/analyzers/annotations/annotations.go (hunk 2 above)
Comment (esnible) [reply galley/pkg/config/analysis/analyzers/annotations/annotations.go]: I thought you would do this under https://github.com/istio/istio/issues/16777 .  If you do not want to write the analyzers I will take a crack at it.

Comment (sushicw) [reply galley/pkg/config/analysis/analyzers/annotations/annotations.go]: I did https://github.com/istio/istio/pull/17770, which covers schema validation. 

It doesn't look like validateAnnotations (in inject.go) gets called anywhere by the validation webhook or the current `istioctl validate`. The issue you referenced doesn't cover this, so we should probably create a new one that does.

**Comment by esnible** [comment]
@geeknoid I need a top level review, just for the _go.mod_ and _go.sum_.  The only module change is for latest istio.io/api to pick up the change you made to _api/annotation/annotations.gen.go_.

**Comment by sushicw** [comment]
@esnible Since this went in I'm seeing this message when I run `istioctl x analyze -k`: 
```
2019-10-15T23:45:31.308862Z	error	source	unable to start watcher for "apps/v1/Deployment": not implemented
```

This seems to be coming from `rt/known.go` [here](https://github.com/istio/istio/pull/17714/commits/1cb9bd0fd1b9ce8138a0420eb99ecf3270d6408d). 

I worry that removing this breaks visibility into deployment resources when using `--use-kube`, at least. Running `istioctl experimental analyze -k --log_output_level analysis:debug,source:debug,processing:debug`, I see that no deployments are in the snapshot:
```
[10] k8s/apps/v1/deployments (@k8s/apps/v1/deployments/0)
```

Re-adding the code block for newInformer, they're there:
```
[10] k8s/apps/v1/deployments (@k8s/apps/v1/deployments/25)
  [0] istio-system/istio-telemetry
  [1] istio-system/prometheus
  [2] kube-system/fluentd-gcp-scaler
  [3] istio-system/istio-policy
  [4] kube-system/metrics-server-v0.3.1
  [5] istio-system/istio-pilot
  [6] istio-system/kiali
  [7] istio-system/istio-tracing
  [8] kube-system/heapster-v1.6.1
  [9] default/productpage-v1
  [10] istio-system/istio-sidecar-injector
  [11] istio-system/grafana
  [12] istio-system/istio-citadel
  [13] default/reviews-v3
  [14] default/reviews-v1
  [15] default/reviews-v2
  [16] kube-system/event-exporter-v0.2.4
  [17] default/ratings-v1
  [18] istio-system/istio-galley
  [19] istio-system/istio-ingressgateway
  [20] kube-system/kube-dns
  [21] default/details-v1
  [22] istio-system/istio-egressgateway
  [23] kube-system/l7-default-backend
  [24] kube-system/kube-dns-autoscaler
```

...but now I get panics in unit test running `make galley-test`. Any idea why?

It looks like we need to iterate on this a bit more. Apologies for not catching this sooner.

**Pull request merged**